src/app/server-cluster/ServerClusterInterface.h

```diff
@@ -0,0 +1,198 @@
+///
+/// Provides metadata as well as interaction processing (attribute read/write and command handling).
+class ServerClusterInterface
+{
+public:
+    ServerClusterInterface();
+    virtual ~ServerClusterInterface() = default;
+
+    // since we use `mNext == this` as a marker for "is in a list"
+    // the assignment of these interfaces is overloaded even for the move operator
+    ServerClusterInterface(ServerClusterInterface && other) :
+        mDataVersion(other.mDataVersion), mNext((other.mNext == &other) ? this : other.mNext)
+    {}
+    ServerClusterInterface(const ServerClusterInterface & other) :
+        mDataVersion(other.mDataVersion), mNext((other.mNext == &other) ? this : other.mNext)
+    {}
+
+    ServerClusterInterface & operator=(const ServerClusterInterface & other)
+    {
+        mDataVersion = other.mDataVersion;
+        mNext        = (other.mNext == &other) ? this : other.mNext;
+        return *this;
+    }
+    ServerClusterInterface & operator=(const ServerClusterInterface && other)
+    {
+        mDataVersion = other.mDataVersion;
+        mNext        = (other.mNext == &other) ? this : other.mNext;
+        return *this;
+    }
+
+    ///////////////////////////////////// Cluster Metadata Support //////////////////////////////////////////////////
+    virtual ClusterId GetClusterId() const = 0;
+
+    // Every cluster must have a data version. Base class implementation to avoid
+    // code duplication
+    //
+    // SPEC - 7.10.3. Cluster Data Version
+    //   A cluster data version is a metadata increment-only counter value, maintained for each cluster instance.
+    //   [...]
+    //   A cluster data version SHALL increment or be set (wrap) to zero
+    //   if incrementing would exceed its maximum value. A cluster data version
+    //   SHALL be maintained for each cluster instance.
+    //   [...]
+    //   A cluster data version SHALL be incremented if any attribute data changes.
+    //
+    DataVersion GetDataVersion() const { return mDataVersion; }
+    void IncreaseDataVersion() { mDataVersion++; }
+
+    /// Cluster flags can be overridden, however most clusters likely have a default of "nothing special".
+    ///
+    /// Default implementation returns a 0/empty quality list.
+    virtual BitFlags<DataModel::ClusterQualityFlags> GetClusterFlags() const;
+
+    ///////////////////////////////////// Attribute Support ////////////////////////////////////////////////////////
+
+    /// ReadAttribute MUST be done on a valid attribute path. `request.path` is expected to have `GetClusterId` as the cluster
+    /// id as well as an attribute that is included in a `Attributes` call.
+    ///
+    /// This MUST HANDLE the following global attributes:
+    ///   - FeatureMap::Id      - generally 0 as a default
+    ///   - ClusterRevision::Id - this is implementation defined
+    ///
+    /// This DOES NOT need to handle attributes built out of cluster metadata. No need to handle:
```

Comment:
> This DOES NOT need to handle attributes built out of cluster metadata

Not sure what that means, or that the average developer would understand what that means. Do you mean "this does NOT handle any implicit ZAP-based data access, such as falling-back on "RAM attributes" supported by legacy ZAP-based flows" ?

Reply:
Will try to rephrase. I meant `this method WILL NOT ever be called with attributeID being those  list attributes` because we handle those inside the interaction model engine. So instead of trying to implement code to build those lists (which takes flash) the interfaces are supposed to implement the metadata calls since we need those anyway (e.g. to check permissions)